lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp

```diff
@@ -2472,4 +2554,85 @@ TEST_F(AMFAppProcedureTest, TestAuthFailureFromSubscribeDbLock) {
   amf_free_ue_context(ue_context_p);
 }
 
+TEST_F(AMFAppProcedureTest, TestPDUSession_LocationUpdateFail) {
+  int rc = RETURNerror;
+  amf_ue_ngap_id_t ue_id = 0;
+  std::vector<MessagesIds> expected_Ids{
+      AMF_APP_NGAP_AMF_UE_ID_NOTIFICATION,  // new registration notification
+                                            // indication to ngap
+      NGAP_NAS_DL_DATA_REQ,                 // Authentication Request to UE
+      NGAP_NAS_DL_DATA_REQ,  // Security Command Mode Request to UE
+      NGAP_NAS_DL_DATA_REQ,  // PDU Session Establishment Request with failure
+                             // sm cause
+      NGAP_NAS_DL_DATA_REQ,  // Deregistaration Accept
+      NGAP_UE_CONTEXT_RELEASE_COMMAND  // UEContextReleaseCommand
+  };
+
+  /* Send the initial UE message */
+  imsi64_t imsi64 = 0;
+  imsi64 = send_initial_ue_message_no_tmsi(amf_app_desc_p, 36, 1, 1, 0, plmn,
+                                           initial_ue_message_hexbuf,
+                                           sizeof(initial_ue_message_hexbuf));
+
+  /* Check if UE Context is created with correct imsi */
+  EXPECT_TRUE(get_ue_id_from_imsi(amf_app_desc_p, imsi64, &ue_id));
+
+  /* Send the authentication response message from subscriberdb */
+  rc = send_proc_authentication_info_answer(imsi, ue_id, true);
+  EXPECT_TRUE(rc == RETURNok);
+
+  /* Send uplink nas message for auth response from UE */
+  rc = send_uplink_nas_message_ue_auth_response(
+      amf_app_desc_p, ue_id, plmn, ue_auth_response_hexbuf,
+      sizeof(ue_auth_response_hexbuf));
+  EXPECT_TRUE(rc == RETURNok);
+
+  /* Send uplink nas message for security mode complete response from UE */
+  rc = send_uplink_nas_message_ue_smc_response(amf_app_desc_p, ue_id, plmn,
+                                               ue_smc_response_hexbuf,
+                                               sizeof(ue_smc_response_hexbuf));
+  EXPECT_TRUE(rc == RETURNok);
+
+  s6a_update_location_ans_t ula_ans = util_amf_send_s6a_ula(imsi);
+  strcpy(ula_ans.imsi, "123456789012345");
+  rc = amf_handle_s6a_update_location_ans(&ula_ans);
+  EXPECT_EQ(rc, RETURNerror);
```

Comment:
On failing to process, s6a-update-location -answer, I think AMF needs to reject the ongoing Registration procedure.

Reply:
@rsarwad To add to what @Chandhu-Wavelabs have mentioned. 
s6a_update_location_ans_t is mainly used for validating the PDU session request w.r.t DNN configuration. This object is not used any where to validate registration request. So we are taking action of sending PDU session reject only when PDU session request is received i.e, if the PDU session request's DNN is not matched with the subscribed configuration from s6a_update_location_ans_t. Similarly we have taken action of sending PDU session reject if s6a response is NULL(Due to any exception from SubscriberDB).

Reply:
Is update location ans used only for DNN verification. Is this  aligned with spec? Because I went through the link, https://www.eventhelix.com/5G/standalone-access-registration/5g-standalone-access-registration.pdf and page number 5, which indicated that after AMF obtains subscriber details, AMF would send Registration Accept message in ICS.

Reply:
Yes you are correct, we are sending Registration Accept only after location information is received. 

In amf_handle_s6a_update_location_ans()
...
  /* FSM takes care of sending registration accept */
  ue_state_handle_message_initial(COMMON_PROCEDURE_INITIATED1,
                                  STATE_EVENT_SEC_MODE_COMPLETE, SESSION_NULL,
                                  ue_mm_context, amf_ctxt_p);
...

The reason why we are not sending registration reject when location information is received invalid/NULL. as we do not have any specific cause here to send (We are not referring this object during registration procedure). When we receive PDU session request and there if we do not have DNN value from location information answer we send PDU session reject with cause "DNN mismatch error"

Reply:
Thanks for explanation.